src/qga.rs

```diff
@@ -45,6 +46,32 @@ impl QgaWrapper {
     /// `sock` is the path to the QGA socket.
     /// `has_kvm` whether or not host supports KVM
     pub fn new(sock: PathBuf, has_kvm: bool) -> Result<Self> {
+        let timeout = if has_kvm {
+            KVM_TIMEOUT
+        } else {
+            EMULATE_TIMEOUT
+        };
+
+        // If we try reading the socket too  early, we'll hang forever and never run the test.
+        // So do the guest_sync first with a timeout to ensure that the VM Guest Agent is up.
+        let end = Instant::now() + timeout;
+        while Instant::now() < end {
+            let qga_stream = match UnixStream::connect(&sock) {
+                Ok(s) => s,
+                Err(e) => {
+                    // If we fail to connect to socket, exit this thread. The main
+                    // thread will detect a hangup and error accordingly.
+                    error!("Failed to connect QGA: {}", e);
+                    return Err(e.into());
+                }
+            };
+            qga_stream.set_read_timeout(Some(Duration::from_secs(1)))?;
+            let mut qga = Qga::from_stream(&qga_stream);
+            let sync_value = &qga_stream as *const _ as usize as i32;
+            if let Ok(_) = qga.guest_sync(sync_value) {
```

Comment:
A little odd but I guess it works. Does the sync value even matter?

Reply:
yeah, guest_sync will validate the sync_value and return an error if it doesn't match.